### PR TITLE
VarifocalNet

### PR DESCRIPTION
**Pull request by hyz-xmaster**
Hi, this is the implementation of our detector,  [VarifocalNet (VFNet)](https://arxiv.org/abs/2008.13367), a dense object detector built on the FCOS.  Please review the code. Thank you.

### CODE REVIEW

File: mmdet/models/dense_heads/vfnet_head.py [hunk 1]
```diff
@@ -0,0 +1,1092 @@
...
+        if self.use_atss:
+            points = torch.stack(
+                (x.reshape(-1), y.reshape(-1)), dim=-1) + \
+                     stride * self.anchor_center_offset
+        else:
+            points = torch.stack(
+                (x.reshape(-1), y.reshape(-1)), dim=-1) + stride // 2
+        return points
+
+    def get_targets(self, points, gt_bboxes_list, gt_labels_list):
```
Comment (yhcao6) [inline mmdet/models/dense_heads/vfnet_head.py:634]: Is this function the same as the get_targets of `FCOSHead`? If it is, you can inherit `FCOSHead` also and call super(self, FCOSHead).get_targets(...)

File: mmdet/models/dense_heads/vfnet_head.py [hunk 2]
```diff
@@ -0,0 +1,1092 @@
...
+            xs, ys = mlvl_points[i][:, 0], mlvl_points[i][:, 1]
+            left = xs - decoded_bboxes[i][..., 0]
+            right = decoded_bboxes[i][..., 2] - xs
+            top = ys - decoded_bboxes[i][..., 1]
+            bottom = decoded_bboxes[i][..., 3] - ys
+            bbox_targets.append(torch.stack((left, top, right, bottom), -1))
+
+        return bbox_targets
+
+    def get_anchors(self, featmap_sizes, img_metas, device='cuda'):
```
Comment (yhcao6) [inline mmdet/models/dense_heads/vfnet_head.py:887]: The following codes are redundant. You can also inherit ATSSHead and call super(self, ATSSHead)

File: mmdet/models/backbones/res2net.py [hunk 3]
```diff
@@ -98,7 +98,7 @@ def __init__(self,
         self.stage_type = stage_type
         self.scales = scales
         self.width = width
-        delattr(self, 'conv2')
+        # delattr(self, 'conv2')
```
Comment (yhcao6) [inline mmdet/models/backbones/res2net.py:101]: May I ask why change this?

File: mmdet/models/losses/varifocal_loss.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+    Args:
+        pred (torch.Tensor): The prediction with shape (N, C), C is the
+            number of classes
+        target (torch.Tensor): The learning target of the iou-aware
+            classification score with shape (N, C), C is the number of classes.
+        weight (torch.Tensor, optional): The weight of loss for each
+                prediction. Defaults to None.
+        alpha (float, optional): A balance factor for Varifocal Loss.
+            Defaults to 0.75.
```
[the remark this answers is not in the document]
Comment (hyz-xmaster) [reply mmdet/models/losses/varifocal_loss.py:26]: will do this

File: mmdet/models/dense_heads/vfnet_head.py [hunk 5]
```diff
@@ -0,0 +1,1092 @@
...
+        return tensor
+    tensor = tensor.clone()
+    dist.all_reduce(tensor, op=dist.ReduceOp.SUM)
+    return tensor
+
+
+@HEADS.register_module()
+class VFNetHead(AnchorFreeHead):
+    """Head of `VarifoclaNet (VFNet): An IoU-aware Dense Object Detector.
+
```
[the remark this answers is not in the document]
Comment (hyz-xmaster) [reply mmdet/models/dense_heads/vfnet_head.py:35]: I have tried, but if I run  pre-commit for correcting the code format, this blank line is automatically added. I will try again.

File: mmdet/models/dense_heads/vfnet_head.py [hunk 6]
```diff
@@ -0,0 +1,1092 @@
...
+    def __init__(self,
+                 num_classes,
+                 in_channels,
+                 regress_ranges=((-1, 64), (64, 128), (128, 256), (256, 512),
+                                 (512, INF)),
+                 center_sampling=False,
+                 center_sample_radius=1.5,
+                 sync_num_pos=True,
+                 gradient_mul=0.1,
+                 bbox_norm_type='reg_denom',
```
[the remark this answers is not in the document]
Comment (hyz-xmaster) [reply mmdet/models/dense_heads/vfnet_head.py:88]: Actually no, denom is the abbreviation of denominator.

File: mmdet/models/backbones/res2net.py (hunk 3 above)
Comment (hyz-xmaster) [reply mmdet/models/backbones/res2net.py:101]: This is the simplest solution for the bug #3714 .

File: mmdet/models/dense_heads/vfnet_head.py [hunk 7]
```diff
@@ -0,0 +1,1092 @@
...
+            stride (int): The corresponding stride for feature maps,
+                used to normalize the bbox prediction when
+                bbox_norm_type = 'stride'.
+            reg_denom (int): The corresponding regression range for feature
+                maps, only used to normalize the bbox prediction when
+                bbox_norm_type = 'reg_denom'.
+
+        Returns:
+            tuple: iou-aware cls scores for each box, bbox predictions and
+            refined bbox predictions of input feature maps.
```
[the remark this answers is not in the document]
Comment (hyz-xmaster) [reply mmdet/models/dense_heads/vfnet_head.py:249]: You mean line 249?

[the remark this answers is not in the document]
Comment (yhcao6) [reply mmdet/models/dense_heads/vfnet_head.py:249]: Yes

File: mmdet/models/backbones/res2net.py (hunk 3 above)
Comment (hyz-xmaster) [reply mmdet/models/backbones/res2net.py]: I have uncommented this line after syncing with the latest updates of MMDetection.

**Comment by hellock** [comment]
ping @yhcao6

**Review by yhcao6 — COMMENTED** [review]
Sry for the late replay. The biggest issue may be the code duplication, for example, `_get_atss_targets` method is almost the same with the `get_target` method in `ATSSHead`, can we inherit ATSSHead and remove the duplication? The similar case happens to the `get_targets` methods in VFNetHead which is almost the same with `get_target` methods of FCOS.

File: mmdet/models/dense_heads/vfnet_head.py [hunk 8]
```diff
@@ -0,0 +1,1092 @@
...
+        # for each image, we compute valid flags of multi level anchors
+        valid_flag_list = []
+        for img_id, img_meta in enumerate(img_metas):
+            multi_level_flags = self.anchor_generator.valid_flags(
+                featmap_sizes, img_meta['pad_shape'], device)
+            valid_flag_list.append(multi_level_flags)
+
+        return anchor_list, valid_flag_list
+
+    def _get_atss_targets(self,
```
[the remark this answers is not in the document]
Comment (hyz-xmaster) [reply mmdet/models/dense_heads/vfnet_head.py:916]: Have tried the multiple inheritance method, but failed due to the class initialization problem. May need help from MMDetection team.

**Comment by hyz-xmaster** [comment]
Hi, @yhcao6, I have finished the pruning and debugging of the code and got it tested. Thank you very much for your example code, which helps me a lot.   Look forward to a further review. Many thanks.

File: mmdet/models/dense_heads/vfnet_head.py [hunk 9]
```diff
@@ -0,0 +1,799 @@
...
+            bottom = decoded_bboxes[i][..., 3] - ys
+            bbox_targets.append(torch.stack((left, top, right, bottom), -1))
+
+        return bbox_targets
+
+    def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+                              missing_keys, unexpected_keys, error_msgs):
+        """Override the method in the parent class to avoid changing para's
+        name."""
+        pass
```
[the remark this answers is not in the document]
Comment (hyz-xmaster) [reply mmdet/models/dense_heads/vfnet_head.py:808]: No need. It works without the `super()._load_from_state_dict()` function, although I do not know the exact reason.

File: mmdet/models/dense_heads/vfnet_head.py [hunk 10]
```diff
@@ -0,0 +1,808 @@
...
+            label_weights.reshape(-1) for label_weights in label_weights_list
+        ]
+        bbox_weights_list = [
+            bbox_weights.reshape(-1) for bbox_weights in bbox_weights_list
+        ]
+        label_weights = torch.cat(label_weights_list)
+        bbox_weights = torch.cat(bbox_weights_list)
+        return labels_list, label_weights, bbox_targets_list, bbox_weights
+
+    def transform_bbox_targets(self, decoded_bboxes, mlvl_points, num_imgs):
```
Comment (yhcao6) [inline mmdet/models/dense_heads/vfnet_head.py:776]: Some codes of this method is duplicate with [`bbox2distance`](https://github.com/open-mmlab/mmdetection/blob/d6b1386711b41c21c009f630ab7b5c011a166af6/mmdet/core/bbox/transforms.py#L143-L164)

**Review by yhcao6 — COMMENTED** [review]
Hi, @hyz-xmaster, This PR is almost done, but still missing the unit tests. Since it adds a new head (VFNetHead) and a new loss (VariFL), it is necessary to add unit tests in [`test_heads.py`](https://github.com/open-mmlab/mmdetection/blob/master/tests/test_models/test_heads.py) and [`test_losses.py`](https://github.com/open-mmlab/mmdetection/blob/master/tests/test_models/test_losses.py).

**Comment by hyz-xmaster** [comment]
> Hi, @hyz-xmaster, This PR is almost done, but still missing the unit tests. Since it adds a new head (VFNetHead) and a new loss (VariFL), it is necessary to add unit tests in [`test_heads.py`](https://github.com/open-mmlab/mmdetection/blob/master/tests/test_models/test_heads.py) and [`test_losses.py`](https://github.com/open-mmlab/mmdetection/blob/master/tests/test_models/test_losses.py).

Hi, @yhcao6, the VFNetHead requires the DeformConv2d function which seems no CPU version. In this case, does the unitest work with the GPU version of DeformConv2d?

**Comment by yhcao6** [comment]
Hi, @hyz-xmaster, in such case you can skip when cuda is not available like [here](https://github.com/open-mmlab/mmdetection/blob/d6b1386711b41c21c009f630ab7b5c011a166af6/tests/test_models/test_heads.py#L295-L296).

**Comment by hyz-xmaster** [comment]
> Hi, @hyz-xmaster, in such case you can skip when cuda is not available like [here](https://github.com/open-mmlab/mmdetection/blob/d6b1386711b41c21c009f630ab7b5c011a166af6/tests/test_models/test_heads.py#L295-L296).

Hi, @yhcao6, thank you. I have noticed that. But when I run this test, it still fails due to the [focal  loss](https://github.com/open-mmlab/mmdetection/blob/d6b1386711b41c21c009f630ab7b5c011a166af6/tests/test_models/test_heads.py#L311) issue.

**Comment by yhcao6** [comment]
> > Hi, @hyz-xmaster, in such case you can skip when cuda is not available like [here](https://github.com/open-mmlab/mmdetection/blob/d6b1386711b41c21c009f630ab7b5c011a166af6/tests/test_models/test_heads.py#L295-L296).
> 
> Hi, @yhcao6, thank you. I have noticed that. But when I run this test, it still fails due to the [focal loss](https://github.com/open-mmlab/mmdetection/blob/d6b1386711b41c21c009f630ab7b5c011a166af6/tests/test_models/test_heads.py#L311) issue.

This should be a bug of `test_fsaf_head_loss` when cuda is available. CI can only cover cpu cases so it doesn't report an error.

File: mmdet/models/dense_heads/vfnet_head.py [hunk 11]
```diff
@@ -0,0 +1,808 @@
...
+
+def get_num_gpus():
+    return int(os.environ['WORLD_SIZE']) if 'WORLD_SIZE' in os.environ else 1
+
+
+def reduce_sum(tensor):
+    if get_num_gpus() <= 1:
+        return tensor
+    tensor = tensor.clone()
+    dist.all_reduce(tensor, op=dist.ReduceOp.SUM)
```
[the remark this answers is not in the document]
Comment (hyz-xmaster) [reply mmdet/models/dense_heads/vfnet_head.py]: Done.

File: mmdet/models/dense_heads/vfnet_head.py (hunk 10 above)
Comment (hyz-xmaster) [reply mmdet/models/dense_heads/vfnet_head.py:772]: simplified with the use of bbox2distance.

**Comment by hyz-xmaster** [comment]
> Hi, @hyz-xmaster, This PR is almost done, but still missing the unit tests. Since it adds a new head (VFNetHead) and a new loss (VariFL), it is necessary to add unit tests in [`test_heads.py`](https://github.com/open-mmlab/mmdetection/blob/master/tests/test_models/test_heads.py) and [`test_losses.py`](https://github.com/open-mmlab/mmdetection/blob/master/tests/test_models/test_losses.py).

The unit tests for VFNetHead and VarifocalLoss are added.

**Review by yhcao6 — APPROVED** [review]

**Review by RyanXLi — APPROVED** [review]

File: configs/vfnet/README.md [hunk 12]
```diff
@@ -0,0 +1,40 @@
+# VarifocalNet: An IoU-aware Dense Object Detector
+
+## Introduction
+**VarifocalNet (VFNet)** learns to predict the IoU-aware classification score which mixes the object presence confidence and localization accuracy together as the detection score for a bounding box. The learning is supervised by the proposed Varifocal Loss (VFL), based on a new star-shaped bounding box feature representation (the features at nine yellow sampling points). Given the new representation, the object localization accuracy is further improved by refining the initially regressed bounding box. The full paper is available at: [https://arxiv.org/abs/2008.13367](https://arxiv.org/abs/2008.13367).
+
+<div align="center">
+  <img src="VFNet.png" width="600px" />
```
Comment (hellock) [inline configs/vfnet/README.md:7]: Replace the image url with `https://user-images.githubusercontent.com/9102141/97464778-4b9ab000-197c-11eb-9283-ab2907ee0252.png` and remove the png from the repo.

Comment (hyz-xmaster) [reply configs/vfnet/README.md]: It's done.

**Comment by hyz-xmaster** [comment]
Thank you very much for your help with this PR, @yhcao6, @RyanXLi and @hellock !

**Comment by hellock** [comment]
@hyz-xmaster Cheers!

**Comment by OpenMMLab-Assistant-007** [comment]
Hi!
@hyz-xmaster 
First of all, we want to express our gratitude for your significant PR in the OpenMMLab project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA

If you have WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤
